lib/scala/cli/src/main/scala/org/enso/cli/Application.scala

```diff
@@ -0,0 +1,266 @@
+
+/**
+  * Represents a CLI application with multiple commands.
+  *
+  * The top-level options can be used to parse global configuration that is
+  * passed to all commands. It can also be used to override behavior (to execute
+  * some other code rather than one of the commands). The top-level options
+  * return a function `() => TopLevelBehaviour[Config]`. When parsed
+  * successfully, that function is executed. It can trigger arbitrary
+  * side-effects. It also defines the further behaviour of the application. If
+  * it returns a [[TopLevelBehavior.Halt]], the application does not parse any
+  * further commands but halts. If it returns [[TopLevelBehavior.Continue]],
+  * parsing continues, trying to execute one of the subcommands. The `Config` is
+  * passed to the subcommands as described below.
+  *
+  * When a command is entered, first its name is matched agains one of the
+  * provided commands. If there is a match, that command is further parsed and
+  * executed. Otherwise, if the name matches one of the related command names,
+  * an error message is returned pointing the user at the right command. If no
+  * related command matches, the plugin manager (if it is provided) is asked to
+  * execute a plugin. If that also fails, an error message is displayed that the
+  * command is not recognized. The message contains similar command names if
+  * available.
+  *
+  * @param commandName default name of the application for use in commands
+  * @param prettyName pretty name of the application for use in text
+  * @param helpHeader short description of the application included at the top
+  *                   of the help text
+  * @param topLevelOpts the top-level options that are used to parse a global
+  *                     config which is passed to every command; can also be
+  *                     used to execute different bahavior than commands; see
+  *                     [[TopLevelBehavior]] for more information
+  * @param commands a sequence of commands supported by the application
+  * @param pluginManager a plugin manager for resolving non-native command
+  *                      extensions
+  * @tparam Config type of configuration that is parsed by the top level
+  *                options and passed to the commands
+  */
+class Application[Config](
+  val commandName: String,
+  val prettyName: String,
+  val helpHeader: String,
+  val topLevelOpts: Opts[() => TopLevelBehavior[Config]],
+  val commands: Seq[Command[Config => Unit]],
+  val pluginManager: Option[PluginManager]
+) {
+
+  /**
+    * A helper overload that accepts the array as provided to the main function.
+    */
+  def run(
+    args: Array[String]
+  ): Either[List[String], Unit] = run(args.toSeq)
+
+  /**
+    * Runs the application logic. Parses the top level options and depending on
+    * its result, possibly runs a command or a plugin.
+    * @return either a list of errors encountered when parsing the options or
```

Comment:
```suggestion
    * its result, possibly runs a command or a plugin.
    *
    * @return either a list of errors encountered when parsing the options or
```
There should always be a blank line between the description and the start of the parameter section. This applies to lots of your comments in this library, so it's worth taking a look at them!